### PR TITLE
Error: Unable to parse elastic URL "http://elasticsearch:9200" and/or PORT "9200" #1442

### PR DESCRIPTION
**Pull request by zg3d**
## Issue This PR Addresses
#1442 

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
Tested the docker build and it was succes full

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/backend/lib/elastic.js [hunk 1]
```diff
@@ -30,9 +30,20 @@ const useMockElastic = process.env.MOCK_ELASTIC;
...
+    return urlObj.port === '' ? `${url}:${port}` : url;
+  }
+
+  return null;
+};
+
 function createElasticClient() {
   try {
-    const elasticUrl = `${ELASTIC_URL}:${ELASTIC_PORT}` || 'http://localhost:9200';
+    const elasticUrl = parseUrl(ELASTIC_URL, ELASTIC_PORT) || 'http://localhost:9200';
```
Comment (manekenpix) [inline src/backend/lib/elastic.js:46]: Take a look at what we do to parse [redis' url](https://github.com/Seneca-CDOT/telescope/blob/7531ea5bbb3eb9f40d16b744e418ead6ca9afb69/src/backend/lib/redis.js#L18), you might be able to do something similar here.

Comment (zg3d) [reply src/backend/lib/elastic.js:46]: @manekenpix  is that not deprecated now. I believe node URL has deprecated it. https://nodejs.org/api/url.html

Comment (humphd) [reply src/backend/lib/elastic.js:46]: Yeah, we can switch to the global `URL` constructor everywhere.  @zg3d want to fix that for Redis while you're doing this surgery?  Also, check if we do it anywhere else.

Comment (manekenpix) [reply src/backend/lib/elastic.js:46]: @zg3d I completely forgot :+1:

Comment (zg3d) [reply src/backend/lib/elastic.js:46]: Ya, I can do that.

File: docker-compose.yml [hunk 2]
```diff
@@ -16,7 +16,7 @@ services:
       - REDIS_URL=redis://redis:${REDIS_PORT}
       # variables for elasticsearch
       - ES_HOST=elasticsearch
-      - ELASTIC_URL=http://elasticsearch:${ELASTIC_PORT}
+      - ELASTIC_URL=http://elasticsearch
```
Comment (humphd) [inline docker-compose.yml:19]: Why this change?  If I want to run es natively, or over the network, why not allow that?

File: src/backend/utils/url-parser.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// parsesUrl to make sure there is only a single port
+const parseUrl = (url, port) => {
+  let result;
```
Comment (humphd) [inline src/backend/utils/url-parser.js:3]: Drop this variable, we'll return early.

File: src/backend/utils/url-parser.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  try {
+    const urlObj = new URL(url);
+    urlObj.port = port;
+    if (urlObj.port !== '') {
```
Comment (humphd) [inline src/backend/utils/url-parser.js:7]: Why bother with this?  Consider:

```js
parseUrl('http://localhost'); // no port (undefined or null), `port = undefined` or `port = null` will both give ''
parseUrl('http://localhost', 3000); // port as number, `port = 3000` will give '3000'
parseUrl('http://localhost', '3000'); // port as string, `port = '3000'` will give '3000'
parseUrl('http://localhost', 'invalid'); // port is invalid, `port = 'invalid'` will give ''
```

Given that `new URL(...).port = ...` won't pollute the port (it will just ignore bad data), shouldn't we always set `urlObj.port = port` and then return `urlObj.href` immediately?

File: src/backend/utils/url-parser.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      result = urlObj.href;
+    }
+  } catch (e) {
+    result = null;
```
Comment (humphd) [inline src/backend/utils/url-parser.js:11]: `return null;`

File: test/url-parser.test.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    expect(result).toBe(expectedReturn);
+  });
+
+  test('urlParser with invalid url', () => {
+    const result = urlParser('', testport);
+    expect(result).toBe(null);
+  });
+
+  test('urlParser with url with no port and no port', () => {
+    const result = urlParser(testUrlWithoutPort, '');
```
Comment (humphd) [inline test/url-parser.test.js:24]: Add test cases for:

- port is undefined
- port is null
- port is a number
- port is a string
- port is an invalid string
- port is a number outside the port range (e.g., 65535 + 1)

**Review by humphd — CHANGES_REQUESTED** [review]
Nice, a few things I see.

File: docker-compose.yml (hunk 2 above)
Comment (zg3d) [reply docker-compose.yml:19]: Was this not the cause of the issue according to @manekenpix

File: src/backend/utils/url-parser.js (hunk 4 above)
Comment (zg3d) [reply src/backend/utils/url-parser.js:7]: is this about the try block ?

Comment (zg3d) [reply src/backend/utils/url-parser.js:7]: @humphd  I don't understand this comment, is this not what I am already doing

Comment (humphd) [reply src/backend/utils/url-parser.js:7]: No, you only return the `.href` in the case that the port is not nothing, meaning you now have 3 different return types:

1. `string` when port is set and we get back `.href`
2. `undefined` when port is `''`
3. `null` when there's a parsing error.

Let's reduce that to 2 and return either `null` or the URL string.  I don't think you need this `if` check, just return `urlObj.ref` in all cases within this try.

**Review by humphd — APPROVED** [review]
Awesome, great work.  One tip on the way you're writing these tests.  You have 6 variables at the top of the describe block that are shared between all the tests.  This isn't a great pattern, and you'd be better off moving all of these into the test functions that use them (maximum isolation between tests, no dependencies, etc).

In your case this isn't going to break, since you used `const` with strings; but if you had non-immutable types (e.g., object references) this could be problematic, if one test updated the data before another used it.

**Comment by humphd** [comment]
Tagging @c3ho since he needs this fix for #1450.

**Comment by c3ho** [comment]
image

I don't know if its me, but I can't get telescope to work with `docker-compose up --build` Telescope keeps crashing on Windows 10

**Comment by humphd** [comment]
@c3ho what's in the log file it generated?

**Comment by c3ho** [comment]
One of the earlier runs:
image

Terminal message in the most recent run:
image

I wiped docker images prior to testing this and still doesn't work. Maybe my elastic container is taking too long to start and Telescope crashes from it?

**Comment by humphd** [comment]
Or maybe this PR isn't parsing the URL in your env correctly?  What are you using for env with these elasitic params?

**Comment by c3ho** [comment]
image

Should it matter though? I thought we're overwriting it with the docker-compose envs when we run `docker-compose up --build`?

**Comment by manekenpix** [comment]
@c3ho change `ELASTIC_DELAY_MS` in your `.env` file, give it more time. The same was happening to me and the problem was that it was taking ES more than `ELASTIC_DELAY_MS` to be ready to store posts.

**Review by manekenpix — APPROVED** [review]

**Review by c3ho — APPROVED** [review]
Got it to work after @manekenpix suggestion.
